library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/InMemoryResourcePack.java

```diff
@@ -0,0 +1,297 @@
+		this.getResourceMap(type).put(id, Suppliers.memoize(resource::get));
+	}
+
+	@Override
+	public @NotNull Future<byte[]> putResourceAsync(@NotNull String fileName,
+			@NotNull Function<@NotNull String, byte @NotNull []> resourceFactory) {
+		Future<byte[]> future = EXECUTOR_SERVICE.submit(() -> resourceFactory.apply(fileName));
+		this.putResource(fileName, () -> {
+			try {
+				return future.get();
+			} catch (InterruptedException | ExecutionException e) {
+				throw new RuntimeException(e);
+			}
+		});
+		return future;
+	}
+
+	@Override
+	public @NotNull Future<byte[]> putResourceAsync(@NotNull ResourceType type, @NotNull Identifier id,
+			@NotNull Function<@NotNull Identifier, byte @NotNull []> resourceFactory) {
+		Future<byte[]> future = EXECUTOR_SERVICE.submit(() -> resourceFactory.apply(id));
+		this.putResource(type, id, () -> {
+			try {
+				return future.get();
+			} catch (InterruptedException | ExecutionException e) {
+				throw new RuntimeException(e);
+			}
+		});
+		return future;
+	}
+
+	@Override
+	public void clearResources(ResourceType type) {
+		this.getResourceMap(type).clear();
+	}
+
+	@Override
+	public void clearResources() {
+		this.root.clear();
+		this.clearResources(ResourceType.CLIENT_RESOURCES);
+		this.clearResources(ResourceType.SERVER_DATA);
+	}
+
+	/**
+	 * Dumps the content of this resource pack into the given path.
+	 *
+	 * @param path the path to dump the resources into
+	 */
+	public void dumpTo(@NotNull Path path) {
+		try {
+			Files.createDirectories(path);
+
+			this.root.forEach((p, resource) -> this.dumpResource(path, p, resource.get()));
+			this.assets.forEach((p, resource) ->
+					this.dumpResource(path, QuiltResourcePack.getResourcePath(ResourceType.CLIENT_RESOURCES, p), resource.get()));
+			this.data.forEach((p, resource) ->
+					this.dumpResource(path, QuiltResourcePack.getResourcePath(ResourceType.SERVER_DATA, p), resource.get()));
+		} catch (IOException e) {
+			LOGGER.error("Failed to write resource pack dump from pack {} to {}.", this.getName(), path, e);
+		}
+	}
+
+	protected void dumpAll() {
+		this.dumpTo(Paths.get("debug", "packs", this.getName()));
+	}
+
+	protected void dumpResource(Path parentPath, String resourcePath, byte[] resource) {
+		try {
+			var p = parentPath.resolve(resourcePath);
+			Files.createDirectories(p.getParent());
+			Files.write(p, resource, StandardOpenOption.CREATE, StandardOpenOption.WRITE,
+					StandardOpenOption.TRUNCATE_EXISTING);
+		} catch (IOException e) {
+			LOGGER.error("Failed to write resource pack dump from pack {}.", this.getName(), e);
+		}
+	}
+
+	private Map<Identifier, Supplier<byte[]>> getResourceMap(ResourceType type) {
+		return switch (type) {
+			case CLIENT_RESOURCES -> this.assets;
+			case SERVER_DATA -> this.data;
+		};
+	}
+
+	static {
+		int threads = Math.max(Runtime.getRuntime().availableProcessors() / 2 - 1, 1);
+		final String property = "quilt.resource_loader.pack.virtual_async_threads";
```

Comment:
This should perhaps be a static field on the top of the file, for better discoverability